### ISSUE TITLE
Validate TS from datasources

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#13574]
### What would you like Renovate to be able to do?

Most datasources receive back some form of structured data to then process. Ideally in such cases we have TS definitions for that data too, and strict type checking. However today, we don't have a way of first validating that the remote data validates against it (e.g. an npm registry's `package.json` conforms to a standard `PackageJson` type).

### If you have any ideas on how this should be implemented, please tell us here.

It would be good if we could have a standard way to do this, e.g.

- Receive data from remote source
- Validate it against our TS definition of that data structure
- Execute our own datasource logic only if it parses

This would save us from embedding validation logic into our datasource logic. Also, having a type for the received data is kind of pointless if everything is `?` and we need to validate every field.

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
Maybe https://github.com/arcanis/typanion or similar can help?

@viceice @JamieMagee @zharinov WDYT?

**Comment by zharinov** [comment]
Once we done with strict null checks, there are variety of nice type-checking tools would be available. Until then, unfortunately they would generate types with all fields as optional.

**Comment by rarkins** [comment]
Is this a duplicate of #9610? I'm not sure if that issue is perhaps too narrow though and that the high level requirement is above

**Comment by zharinov** [comment]
Yes, but datasources would be a really good start for the long journey
